### PR TITLE
Making $<TARGET_FILE:some_target> work for non imported targets

### PR DESCRIPTION
**Pull request by Volker-Weissmann**
This PR makes that the generator expression $<TARGET_FILE:some_target> gets correctly translated to /path/to/some/target, even if some_target is not imported, i.e. the cmake command `add_executable` was called without the IMPORTED argument...

Note: [This other PR from me](https://github.com/mesonbuild/meson/pull/10545) made an error message nicer, and this PR here makes that this error does not occur (at least in the cases I tested for).

### CODE REVIEW

**Comment by Volker-Weissmann** [comment]
The tests failed, because what I wrote does not work when cross-compiling. I can reproduce it locally and will fix it.

File: mesonbuild/cmake/traceparser.py [hunk 1]
```diff
@@ -354,16 +357,16 @@ def _cmake_add_executable(self, tline: CMakeTraceLine) -> None:
...
-        self.targets[args[0]] = CMakeTarget(args[0], 'EXECUTABLE', {}, tline=tline, imported=is_imported)
+        if 'IMPORTED' in args:
+            args.remove('IMPORTED')
+            if len(args) < 1:
+                return self._gen_exception('add_executable', 'requires at least 1 argument', tline)
+            self.targets[args[0]] = CMakeTarget(args[0], None, 'EXECUTABLE', {}, tline=tline, imported=True)
+        else:
+            if len(args) < 1:
+                return self._gen_exception('add_executable', 'requires at least 1 argument', tline)
+            build_path = self.build_dir.parent / args[0]
```
Comment (mensinda) [inline mesonbuild/cmake/traceparser.py:368]: This looks a bit too fragile to me (and is probably also the cause of the current test failures). Ideally, we could use [`Executable.get_filename()`](https://github.com/mesonbuild/meson/blob/master/mesonbuild/build.py#L1296) here, but constructing a new `Executable` looks too expensive for something that won't be needed in 90% of all cases.

The best solution I can come up with right now is to defer the construction of the `build_path` by passing a lambda instead to `CMakeTarget` and have it lazily evaluate (and cache) the result if required.

**Review by mensinda — CHANGES_REQUESTED** [review]
I am happy with this change, except for the one comment and the red CI.

**Comment by Volker-Weissmann** [comment]
There is a deeper problem:
If you are crosscompiling, you need to build all generators for the build machine, not the host machine. In my test case, I used
```
add_executable(mygen gen.c)
```
to build this generator. This is [broken, even if you use just cmake instead of cmake+meson](https://cmake.org/cmake/help/book/mastering-cmake/chapter/Cross%20Compiling%20With%20CMake.html#running-executables-built-in-the-project). Meson could try to fix this by checking which executable is used as a generator, then build those for the build machine and use them. But this would be a lot of work for something that isn't even supported by cmake itself. So I'm just going to exclude my test from the cross-compiling tests. They failed because ninja tried to execute an arm binary on an x86 machine. (Build machine was x86, host machine was arm.)

**Comment by Volker-Weissmann** [comment]
My plan to fix this is as follows (and no, I don't know what I'm doing):

Make `parse_generator_expressions`  return an object of a new class, `CmakeGenExpr`. This class will have a method called `as_string`.
If TARGET_NAME or similar occurs in the generator expression this method will terminate meson with a message saying "missing feature in meson you cannot build this"
It TARGET_NAME or similar things do not occur in the generator expression, `as_string` will return what `parse_generator_expressions` returns now.

Then I put `as_string()` calls everywhere the result of `parse_generator_expressions` is currently used with the exception of `pretend_to_be_meson`.

`pretend_to_be_meson` currently returns, among other things, something like
```
custom_tgt(
    ...,
    command: ['something', 'something', 'str returned by parse_generator_expressions']
)
```
In other words, it puts a `StringNode(Token('string', some/path, 0, 0, 0, None, 'str returned by parser_generator_expressions'))` in its output.  If this StringNode would be replaced by the result of `as_meson_ast`, a new method of `CmakeGenExpr`, then the ast that pretend_to_be_meson generates could look like this
```
custom_tgt(
    ...,
    command: ['something', 'something', some_target.full_path()]
)
```

This should be a robust and extensible solution. The only cause for future problems is that the "missing feature in meson you cannot build this" might occur.

**Comment by Volker-Weissmann** [comment]
The failing tests also fail on the master branch.
It's ready to merge.

**Comment by Volker-Weissmann** [comment]
Will this get merged eventually?

**Comment by eli-schwartz** [comment]
@mensinda ping?

File: mesonbuild/cmake/generator.py [hunk 2]
```diff
@@ -67,8 +66,7 @@ def target_property(arg: str) -> str:
 
     def target_file(arg: str) -> str:
         if arg not in trace.targets:
-            mlog.warning(f"Unable to evaluate the cmake variable '$<TARGET_FILE:{arg}>'.")
-            return ''
+            return arg
```
Comment (mensinda) [inline mesonbuild/cmake/generator.py:69]: Technically, just returning arg here is wrong. (Since it will return `cm_exe_local` instead of the actual location of the newly built exe). This works in this case, since `add_custom_command` also accepts a target name as the first argument.

So, while returning `arg` is technically wrong, I would argue that it is "less wrong" than returning an empty string, since it works in *some* cases (but will still potentially fail for others).

However, the warning should not be removed, since we are still guessing and future issues will be easier to identify (mayebe reword it to `Unable to evaluate the...` --> `Unable to correctly evaluate the...`).

**Review by mensinda — COMMENTED** [review]
I have only one minor issue. Once this is resolved, this PR is ready to merge (assuming the CI is green).

Sorry for the late review.

File: test cases/cmake/26 target file/meson.build [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+# gen_src = custom_target('gen-output',
+#                         output : ['out.c'],
+#                         command : [volker_inter_exe, '@OUTPUT@'], install: true, install_dir: '.')
+
+# exe = executable('volker_exe', gen_src)
+x = 'abc' + 'def' + 'ghi'
+
+if meson.is_cross_build()
+    message('Test skipped, see https://github.com/mesonbuild/meson/pull/10607#issuecomment-1188117276 for an explanation.')
```
Comment (eli-schwartz) [inline test cases/cmake/26 target file/meson.build:24]: This should be an `error('MESON_SKIP_TEST: explanation here')` probably? The test harness will then log it as `[SKIPPED]` instead of `[SUCCESS]`.

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

File: mesonbuild/cmake/generator.py (hunk 2 above)
Comment (Volker-Weissmann) [reply mesonbuild/cmake/generator.py:70]: Done

File: test cases/cmake/26 target file/meson.build (hunk 3 above)
Comment (Volker-Weissmann) [reply test cases/cmake/26 target file/meson.build]: Done

**Comment by mensinda** [comment]
The vs2019 test failures look relevant: https://dev.azure.com/jussi0947/jussi/_build/results?buildId=16963&view=results

**Comment by Volker-Weissmann** [comment]
> The vs2019 test failures look relevant: https://dev.azure.com/jussi0947/jussi/_build/results?buildId=16963&view=results

Yes it does. 
I think we pause this PR until I find the inner strength to work on windows.
